presto-spi/src/main/java/com/facebook/presto/spi/prerequisites/QueryPrerequisites.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.prerequisites;
+
+import com.facebook.presto.spi.QueryId;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An interface to plugin custom business logic that will be executed before the query is queued.
+ */
+public interface QueryPrerequisites
+{
+    /**
+     * Given the query context, implementations can perform actions or ensure that all conditions
+     * are ready for the query to execute. The returned <code>CompletableFuture</code> will indicate
+     * when the query is ready to be queued for execution. If the returned future finishes successfully,
+     * it will trigger the query to be queued and its failure will fail the query.
+     */
+    CompletableFuture<?> waitForPrerequisites(QueryId queryId, QueryPrerequisitesContext context);
+
+    /**
+     * Optional method for the implementations to implement if they want to be informed about the finishing
+     * of queries (either successfully or unsuccessfully).
+     */
+    default void queryFinished(QueryId queryId)
+    {
+    }
```

Comment:
doesn't feel like the right place to add logic which can be executed **after** the query is finished running since this is for just for query prerequisites.

Reply:
This is to inform the QueryPrerequisites that a query has finished and they can cleanup state if needed. This is useful since the query may fail before the prerequisites are satisfied (example - a user may kill their query).

Can you clarify your concerns here?

Reply:
The method signature read like actions to be done **post** the query finished executing. Since this class is for processing prerequisites before the query run, mixing **pre** and **post** actions was my concern.

Reply:
I somehow don't have the same concern and hence unable to think of a solution. Do you have suggestions on how what I mentioned can be achieved?

Reply:
Having a separate class called PostQueryActions and add methods for  queryFinished there. Since this is a high pri PR don't get blocked on this. Keep it in the backlog and see if this is possible.